emu/src/memory/internal_memory.rs

```diff
@@ -209,19 +214,20 @@ impl InternalMemory {
 }
 
 impl IoDevice for InternalMemory {
-    type Address = u32;
+    type Address = usize;
     type Value = u8;
 
     fn read_at(&self, address: Self::Address) -> Self::Value {
         match address {
-            0x00000000..=0x00003FFF => self.bios_system_rom[(address) as usize],
-            0x02000000..=0x0203FFFF => self.working_ram[(address - 0x02000000) as usize],
-            0x03000000..=0x03007FFF => self.working_iram[(address - 0x03000000) as usize],
+            0x00000000..=0x00003FFF => self.bios_system_rom[address],
+            0x02000000..=0x0203FFFF => self.working_ram[address - 0x02000000],
+            0x03000000..=0x03007FFF => self.working_iram[address - 0x03000000],
             0x04000000..=0x04000055 => self.read_address_lcd_register(address),
             0x04000100..=0x0400010E => self.read_address_timer_register(address),
-            0x05000000..=0x050001FF => self.bg_palette_ram[(address - 0x05000000) as usize],
-            0x05000200..=0x050003FF => self.obj_palette_ram[(address - 0x05000200) as usize],
-            0x06000000..=0x06017FFF => self.video_ram[(address - 0x06000000) as usize],
+            0x05000000..=0x050001FF => self.bg_palette_ram[address - 0x05000000],
+            0x05000200..=0x050003FF => self.obj_palette_ram[address - 0x05000200],
+            0x06000000..=0x06017FFF => self.video_ram[address - 0x06000000],
+            0x08000000..=0x0FFFFFFF => self.rom[address - 0x08000000],
```

Comment:
I think this should be splitted in multiple lines. Accordin to gbatek:

> 08000000-09FFFFFF   Game Pak ROM/FlashROM (max 32MB) - Wait State 0
  0A000000-0BFFFFFF   Game Pak ROM/FlashROM (max 32MB) - Wait State 1
  0C000000-0DFFFFFF   Game Pak ROM/FlashROM (max 32MB) - Wait State 2
  0E000000-0E00FFFF   Game Pak SRAM    (max 64 KBytes) - 8bit Bus width
  0E010000-0FFFFFFF   Not used

So I would do something like:

```rust
0x08000000..=0x09FFFFFF => self.rom[address - 0x08000000],
0x0A000000..=0x0BFFFFFF => self.rom[address - 0x0A000000],
0x0C000000..=0x0DFFFFFF => self.rom[address - 0x0C000000],
0x0E000000..=0x0E00FFFF  => self.rom_sram[address - 0x0E000000],
```

And then `rom_sram` should also be implemented in the `write_at` part since it's a ram

Reply:
Add a comment for don't forget